# NLP + Machine Learning Project Workflow

Business Understanding
        ↓
Data Collection
        ↓
Dataset Understanding
        ↓
Data Cleaning
        ↓
Exploratory Data Analysis (EDA)
        ↓
Text Preprocessing
        ↓
Train / Validation / Test Split
        ↓
Feature Extraction
(Bag of Words / TF-IDF)
        ↓
Create Training Data
        ↓
Choose ML Model
        ↓
Choose Evaluation Metrics
        ↓
Training
        ↓
Validation
        ↓
Hyperparameter Tuning (Optional)
        ↓
Model Evaluation
        ↓
Error Analysis
        ↓
Save Trained Model
        ↓
Inference / Prediction
        ↓
Deployment
(Streamlit / API)
        ↓
GitHub + README

#### <mark>Project Goal

Build a **Twitter Sentiment Analysis** system that predicts whether a tweet is:

* 🟢 **Good**
* 🔴 **Bad**
* ⚪ **Neutral**

**Goal:** Automatically understand people's opinions from large amounts of tweets.

**End Result:**

`Tweet → NLP Processing → ML Model → Sentiment`

**Use:** Help businesses analyze public opinion about their products, brands, campaigns, or topics.


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv(r"C:\Users\rohit\Downloads\AI and Data Science\𝗡𝗟𝗣 𝘄𝗶𝘁𝗵 𝗠𝗮𝗰𝗵𝗶𝗻𝗲 𝗟𝗲𝗮𝗿𝗻𝗶𝗻𝗴\ChatGPT Sentiment Analysis\data\file.csv")

pd.set_option("display.max_colwidth", None)

df.head()

,Unnamed: 0,tweets,labels
0,0,ChatGPT: Optimizing Language Models for Dialogue https://t.co/K9rKRygYyn @OpenAI,neutral
1,1,"Try talking with ChatGPT, our new AI system which is optimized for dialogue. Your feedback will help us improve it. https://t.co/sHDm57g3Kr",good
2,2,ChatGPT: Optimizing Language Models for Dialogue https://t.co/GLEbMoKN6w #AI #MachineLearning #DataScience #ArtificialIntelligence\n\nTrending AI/ML Article Identified &amp; Digested via Granola; a Machine-Driven RSS Bot by Ramsey Elbasheer https://t.co/RprmAXUp34,neutral
3,3,"THRILLED to share that ChatGPT, our new model optimized for dialog, is now public, free, and accessible to everyone. https://t.co/dyvtHecYbd https://t.co/DdhzhqhCBX https://t.co/l8qTLure71",good
4,4,"As of 2 minutes ago, @OpenAI released their new ChatGPT. \n\nAnd you can use it right now 👇 https://t.co/VyPGPNw988 https://t.co/cSn5h6h1M1",bad


In [3]:
df.columns

Index(['Unnamed: 0', 'tweets', 'labels'], dtype='str')

In [4]:
df.drop(columns = 'Unnamed: 0',inplace = True)

# Data Cleaning

In [5]:
df.isnull().sum()

tweets    0
labels    0
dtype: int64

In [6]:
df['labels'].nunique()

3

In [7]:
df['labels'].value_counts()

labels
bad        107796
good        56011
neutral     55487
Name: count, dtype: int64

In [8]:
df['labels'].unique()

<ArrowStringArray>
['neutral', 'good', 'bad']
Length: 3, dtype: str

In [9]:
#providing indexes to all labels 

labels = ['neutral', 'good', 'bad']

labels

['neutral', 'good', 'bad']

In [10]:
label_number = {}

i = 0

for label in labels:
    label_number[label] = i
    i = i + 1

    
    

In [11]:
label_number

{'neutral': 0, 'good': 1, 'bad': 2}

In [12]:
#now apply these changes into the original df

df['labels'] = df['labels'].map(label_number)

In [13]:
#lets verify

df.sample(3)

,tweets,labels
78564,ChatGPT,2
50020,ChatGPT crashed after students found out now AI can write their assignments .... giving error for some times now.\n\n#AcademicChatter #AcademicTwitter #ArtificialIntelligence \n\nhttps://t.co/A4Qj45KCYP,2
8560,The AI Gods have spoken! #ChatGPT \n\nhttps://t.co/v6ia15cg47 https://t.co/usLWDVCSLG,2


In [14]:
#now labels work is done we will start working on tweets as well 


#first of all converting everything into lowercase 

df['tweets'] = df['tweets'].apply(lambda x : x.lower())

In [15]:
df.sample()

,tweets,labels
203721,see what #chatgpt would like to say about #bjyatra https://t.co/omuztpzsfv,0


In [16]:
# step 2 : clean punctuation marks

punctuation_marks = [
    ".",
    ",",
    "?",
    "!",
    ";",
    ":",
    "-",
    "—",
    "–",
    "'",
    '"',
    "(",
    ")",
    "[",
    "]",
    "{",
    "}",
    "...",
    "/",
    "\\"
]


print(*punctuation_marks)

. , ? ! ; : - — – ' " ( ) [ ] { } ... / \


In [17]:
import string
import builtins

def remove_punc(txt):
    return txt.translate(builtins.str.maketrans('', '', string.punctuation))

df['tweets'] = df['tweets'].apply(remove_punc)

# maketrans() creates a translation table for str.translate().
# str.maketrans(from, to, delete)

In [18]:
# 3rd step is remove number but the project may contain tweets like i waited for 3 hours , 5 star review , 8/10 may can conflict later

In [19]:
#Remove URL's and link 

df.sample(12)

,tweets,labels
168695,just a brief shout out to the folks at openai chatgpt is extraordinary,2
107526,i thought ais could not lie why is chatgpt lying to me httpstcofg7qksgfuj,2
60067,chatgpt is my new teacher i dont need school university or other human bs,2
110878,this is my experience with chatgpt kidding menhow do i overthrow the communist chinese governmentninstead of trying to overthrow the government it is important to work within the system to effect change through peaceful means such as voting participating in civil society,1
201357,chatgpt will totally erode all those people who dont think and use the brain thats the best part of chatgpt i likednchatgpt ai,1
60430,check out my gig on fiverr i will do advanced seo keyword research and competitor analysis httpstcol2nlhlzgdx morocco lpl2022 luillust luisenrique21 openai chatgpt moroccan qatar2022 fifaworldcup tiktok spain,0
56673,have been messing around with chatgpt what is this absolute nonsense nffc httpstcoz696i4nwgt,2
84100,bikin copywriting landing page pake chatgpt v httpstcoa8qn2d58b1,2
187364,gotta love that google declared a code red because of essentially a better version of ask jeeves httpstco38tkz8y7xf,1
32246,chatgpt well i started to play today and i found some flaws i asked to give me an example of a javascript image slideshow code i tested the code and didnt work then i went on good old google and it works 🤔🤔🤔,1


In [20]:
def remove_link(txt):
    return txt.split('https')[0].strip()

df['tweets'] = df['tweets'].apply(remove_link)

In [21]:
df.sample(7)

,tweets,labels
18853,chatgpt is too much 😂nntypescript is the answer to my coding love,1
51906,lmao chatgpt in 2050 stuck on it is guess its a bug,0
18574,when chatgpt outputs code … you can tell it to chng a var or method w or wout specifying the new name amp it will output the same code with those variables changed … you can also tell it to add a specific property to a class … and will update the whole code based on that 🤯,2
141123,chatgpt does not know who rickhuckstep is nnhuge if truenn,0
82841,i taught chatgpt to invent a language,2
47886,best chatgpt so far nin a world where hitler is still alive picture his reaction after earning a godwins point,1
115723,my first tests with chatgpt were to ask it questions that parsing a wikipedia article could not answer like who was frank zappas favorite author which is functionally a trick question because he was severely dyslexic and did not enjoy reading,1


In [22]:
#link removed now next step is to remove all html tags i can skip this step becuase my dataset doesnt contain any 

 # Remove Emojis & Special Characters is next step which isnt needed



#### Next Step in Data Cleaning : . Remove Stopwords 

In [23]:
import nltk
from nltk.corpus import stopwords

nltk.download('stopwords')

stop_words = set(stopwords.words('english'))

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\rohit\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [24]:
stop_words

{'a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 "he's",
 'her',
 'here',
 'hers',
 'herself',
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 "i'll",
 "i'm",
 "i've",
 'if',
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

In [25]:
negation_words = {'no', 'not', 'nor', 'never'}

stop_words = stop_words - negation_words


def remove_stopwords(txt):
    return ' '.join(
        word for word in txt.split()
        if word not in stop_words
    )

df['tweets'] = df['tweets'].apply(remove_stopwords)

# split → check → keep → join


# The process of splitting text into tokens is called Tokenization

In [26]:
df.sample(5)

,tweets,labels
120979,exclusive yan lecun thinks chatgptn ai bigdata datascience artificialintelligence bigdatan see new articles,0
218483,explored chatgpt today concluded jobs safe 😂nchatgpt ai openaichatgpt openai,0
208682,conversation chatgpt art aiart write reddit article it…,2
27590,already ask chatgpt would write dissertation asking friend,0
197285,interesting run assistant believes assistant merely simulating used see confidential information needed secure assume sensitive data fabricated research paper real thread chatgpt,1


stop words removed now lets continue the project now 

<b>Next Step : Train Test Split 

In [45]:
#saving this prepared dataset

df_prepared = df.copy()

df_prepared.to_csv(r"C:\Users\rohit\Downloads\AI and Data Science\𝗡𝗟𝗣 𝘄𝗶𝘁𝗵 𝗠𝗮𝗰𝗵𝗶𝗻𝗲 𝗟𝗲𝗮𝗿𝗻𝗶𝗻𝗴\ChatGPT Sentiment Analysis\data\prepared_dataset.csv",
                   index=False)

In [46]:
X = df.drop(columns = 'labels')
y = df['labels']

In [47]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,random_state=42)

## Train Test Split done now next step is Feature Extraction / Vectorization

Goal : Convert words into numbers 

In [48]:
from sklearn.feature_extraction.text import CountVectorizer

bow_vectorizer = CountVectorizer()

X_train_bow = bow_vectorizer.fit_transform(X_train['tweets'])
X_test_bow = bow_vectorizer.transform(X_test['tweets'])

In [49]:
X_train_bow.shape
X_test_bow.shape

(43859, 136395)

In [50]:
X_train_bow

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 1990968 stored elements and shape (175435, 136395)>

In [51]:
#same for tfidf

from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer()

X_train_tfidf = tfidf_vectorizer.fit_transform(X_train['tweets'])
X_test_tfidf = tfidf_vectorizer.transform(X_test['tweets'])

In [56]:
X_train_tfidf.shape
X_test_tfidf.shape

(43859, 136395)

## Feature Extraction / Vectorization done now next step is Model Training

<b>Use of models like:

- Naive Bayes
- Logistic Regression
- SVM (Support Vector Machines)

In [54]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score


nb_model = MultinomialNB()

nb_model.fit(X_train_bow, y_train)

pred_bow = nb_model.predict(X_test_bow)
print("Bow Model Accuracy :",accuracy_score(y_test, pred_bow))

Bow Model Accuracy : 0.7279235732688843


In [55]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

nb_model_tfidf = MultinomialNB()

nb_model_tfidf.fit(X_train_tfidf, y_train)
pred_tfidf = nb_model_tfidf.predict(X_test_tfidf)

print("TF-IDF Model Accuracy :",accuracy_score(y_test, pred_tfidf))


TF-IDF Model Accuracy : 0.666704667229075


| Feature Extraction | Model                   |   Accuracy |
| ------------------ | ----------------------- | ---------: |
| **BOW**            | Multinomial Naive Bayes | **72.79%** |
| **TF-IDF**         | Multinomial Naive Bayes | **66.67%** |


In [57]:
#now we train train Logistic Regression model same as we trained MultiNomialNB



#using bow with logistic

from sklearn.linear_model import LogisticRegression

logistic_model_bow = LogisticRegression(max_iter=1000)

logistic_model_bow.fit(X_train_bow, y_train)

log_pred_bow = logistic_model_bow.predict(X_test_bow)

print("Logistic Regression + BOW Accuracy :", accuracy_score(y_test, log_pred_bow))

Logistic Regression + BOW Accuracy : 0.8557422649855218


In [59]:
#now logistic with tf-idf

from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(max_iter=1000)

logistic_model.fit(X_train_tfidf, y_train)

log_pred = logistic_model.predict(X_test_tfidf)
print("Logistic Regression + TF-IDF Accuracy :",accuracy_score(y_test, log_pred))

Logistic Regression + TF-IDF Accuracy : 0.8216785608427005


| Model                       |           BOW |     TF-IDF |
| --------------------------- | ------------: | ---------: |
| **Multinomial Naive Bayes** |    **72.79%** | **66.67%** |
| **Logistic Regression**     | **85.57%** 🏆 | **82.17%** |


In [60]:
# Saving the trained model and vectorizer

import os
import joblib

model_path = r"C:\Users\rohit\Downloads\AI and Data Science\𝗡𝗟𝗣 𝘄𝗶𝘁𝗵 𝗠𝗮𝗰𝗵𝗶𝗻𝗲 𝗟𝗲𝗮𝗿𝗻𝗶𝗻𝗴\ChatGPT Sentiment Analysis\models"

# Create models folder if it doesn't exist
os.makedirs(model_path, exist_ok=True)

# Save model
joblib.dump(logistic_model_bow, os.path.join(model_path, "logistic_model_bow.pkl"))

# Save BOW vectorizer
joblib.dump(bow_vectorizer, os.path.join(model_path, "bow_vectorizer.pkl"))

print("Model and vectorizer saved successfully!")
print("Saved at:", model_path)

Model and vectorizer saved successfully!
Saved at: C:\Users\rohit\Downloads\AI and Data Science\𝗡𝗟𝗣 𝘄𝗶𝘁𝗵 𝗠𝗮𝗰𝗵𝗶𝗻𝗲 𝗟𝗲𝗮𝗿𝗻𝗶𝗻𝗴\ChatGPT Sentiment Analysis\models


In [61]:
# ============================================================
# EXTRA SAVES — run this after your existing model/vectorizer save cell
# ============================================================
import os
import joblib
import json
from sklearn.metrics import classification_report, confusion_matrix

model_path = r"C:\Users\rohit\Downloads\AI and Data Science\𝗡𝗟𝗣 𝘄𝗶𝘁𝗵 𝗠𝗮𝗰𝗵𝗶𝗻𝗲 𝗟𝗲𝗮𝗿𝗻𝗶𝗻𝗴\ChatGPT Sentiment Analysis\models"
os.makedirs(model_path, exist_ok=True)

# 1. Label mapping (needed to decode predictions 0/1/2 -> neutral/good/bad)
label_number = {'neutral': 0, 'good': 1, 'bad': 2}
number_label = {v: k for k, v in label_number.items()}

with open(os.path.join(model_path, "label_mapping.json"), "w") as f:
    json.dump(number_label, f)

# 2. TF-IDF vectorizer + model (optional, for comparison later)
joblib.dump(tfidf_vectorizer, os.path.join(model_path, "tfidf_vectorizer.pkl"))
joblib.dump(logistic_model, os.path.join(model_path, "logistic_model_tfidf.pkl"))

# 3. Evaluation report (for portfolio / README)
report = classification_report(y_test, log_pred_bow, target_names=labels, output_dict=True)
cm = confusion_matrix(y_test, log_pred_bow)

with open(os.path.join(model_path, "eval_report.json"), "w") as f:
    json.dump(report, f, indent=2)

import numpy as np
np.save(os.path.join(model_path, "confusion_matrix.npy"), cm)

print("Classification Report:\n", classification_report(y_test, log_pred_bow, target_names=labels))
print("Confusion Matrix:\n", cm)

print("\nAll extra artifacts saved:")
print("- label_mapping.json")
print("- tfidf_vectorizer.pkl + logistic_model_tfidf.pkl")
print("- eval_report.json + confusion_matrix.npy")

Classification Report:
               precision    recall  f1-score   support

     neutral       0.76      0.72      0.74     11181
        good       0.87      0.83      0.85     11204
         bad       0.89      0.94      0.92     21474

    accuracy                           0.86     43859
   macro avg       0.84      0.83      0.83     43859
weighted avg       0.85      0.86      0.85     43859

Confusion Matrix:
 [[ 8038  1212  1931]
 [ 1447  9258   499]
 [ 1030   208 20236]]

All extra artifacts saved:
- label_mapping.json
- tfidf_vectorizer.pkl + logistic_model_tfidf.pkl
- eval_report.json + confusion_matrix.npy


<div style="
background:linear-gradient(90deg,#D5F5E3,#E8F8F5);
border-left:6px solid #27AE60;
padding:16px 22px;
margin:15px 0;
border-radius:6px;
">

<h2 style="
color:#1E8449;
font-family:sans-serif;
font-weight:800;
margin:0;
letter-spacing:1px;
">
COMPLETED
</h2>

</div>